### DISCUSSION TITLE
How can one add a value to an IntVar domain?

### DISCUSSION THREAD

**Discussion by jmarca** [google/or-tools#3284] [Routing (and legacy CP) questions] [answered]
OR Tools 9.3, calling from Python, using the routing solver

So in my code, I remove a value from the domain of an IntVar, but then later, under certain circumstances, I want to put that variable back into the domain.

Essentially, I want to perform the opposite of 

```
routing.NextVar(idxA).RemoveValue(idxB)
```

I tried to use "SetValues()" but that did not work.  And I didn't have high hopes, given that the docs say SetValues makes the domain the *intersection* of the current range and the incoming range.

For what its worth, what I tried was:

```
    domain = {d for d in self.routing.NextVar(midx).DomainIteratorAux(False)}
    domain.add(nidx)
    self.routing.NextVar(midx).SetValues(domain)
```
Since I see no such operator anywhere, perhaps there is a good reason for its omission.  If it is not an oversight and was omitted on purpose, I would appreciate an explanation of why it is a bad idea to insert values into the range of an IntVar.

**Comment by lperron** [comment, marked as answer]
You can't. All operations are coded to be monotonic.

**Comment by jmarca** [comment]
Okay thanks for the clarification.  The domain can shrink, but never grow.

I don't really need it I guess; I just have to think ahead before removing the
indices from the NextVar().

On Thu, May 05, 2022 at 09:03:35PM -0700, Laurent Perron wrote:
> You can't. All operations are coded to be monotonic.
>
> --
> Reply to this email directly or view it on GitHub:
> https://github.com/google/or-tools/discussions/3284#discussioncomment-2698142
> You are receiving this because you authored the thread.
>
> Message ID: ***@***.***>

--

James E. Marca
Activimetrics LLC
